File: Data Structures & Algorithms/valid-tree/submission-7.py
class Solution:
    def validTree(self, n: int, edges: list[list[int]]) -> bool:
        adjs = {vertex: set() for vertex in range(n)}
        for u, v in edges:
            if u == v:
                return False
            adjs[u].add(v)
            adjs[v].add(u)

        path = set()
        visited = set()

        def dfs(vertex, prev_vertex):
            if vertex in path:
                return False

            path.add(vertex)
            visited.add(vertex)

            for adj_vertex in adjs[vertex]:
                if (
                    adj_vertex != prev_vertex and 
                    not dfs(adj_vertex, vertex)
                ):
                    return False
            
            path.remove(vertex)
            return True
        
        return dfs(0, -1) and len(visited) == n

File: Data Structures & Algorithms/count-connected-components/submission-7.js
class Solution {
   /**
    * Time complexity: O(V + E)
    * Auxiliary space complexity: O(V + E)
    * Tags: dfs, recursion, graph, dsu
    * BFS
    * @param {number} nodeCount
    * @param {number[][]} edges
    * @return {}
    */
   countComponents(nodeCount, edges) {
      const adjs = new Map();
      for (let node = 0; node < nodeCount; node++) {
         adjs.set(node, new Set());
      }
      for (const [u, v] of edges) {
         adjs.set(u, adjs.get(u).add(v));
         adjs.set(v, adjs.get(v).add(u));
      }

      const visited = new Set();
      let components = 0;

      const bfs = (node) => {
         const queue = new Queue([node]);

         while (!queue.isEmpty()) {
            const node = queue.dequeue();
            visited.add(node);
            for (const adjNode of adjs.get(node))
               if (!visited.has(adjNode))
                  queue.enqueue(adjNode);
         }
      };
      for (let node = 0; node < nodeCount; node++) {
         if (!visited.has(node)) {
            bfs(node)
            components++;
         }
      }
      return components
   };
}